### ISSUE TITLE
Gatsby type inference not working when creating a union

### ISSUE THREAD

**Issue by jeffsee55** [gatsbyjs/gatsby#19355]
## Summary
I need to define a union type for my frontmatter data, but when I do that I lose Gatsby's type inference. I'm not sure if the title of this issue is accurate - it might the result of the way I'm approaching this and might not have anything to do with the union I'm creating. I just know that in the docs it's clear that gatsby is supposed to infer types unless explicitly told otherwise.

## Relevant Information

I'm trying to work with a gatsby site using gatsby-transformer-remark and gatsby-source-filesystem but I want to do something a little bit special with my frontmatter (I plan to turn this into a plugin). I have a section in the frontmatter called blocks which takes an array of 'blocks' - however there are different fields on each block - specified by the template in the given block:
```md
# some-file.md
---
title: First Blog
blocks: 
- template: block__hero
  heroField1: '## Hello world'
  heroField2: '## Hello world'
- template: block__columns
  columnField1: '## Hello again'
  columnField2: '## Hello again'
---
```

As you can see `block__hero` and  `block__columns` are two completely different data structures and I'd like to treat them as such in my graphql schema. Right now I get the following

```gql
query MyQuery {
  allMarkdownRemark {
    edges {
      node {
        frontmatter {
          title
          blocks {
            template
            heroField1
            heroField2
            columnField1
            columnField2
          }
        }
      }
    }
  }
}
```

Since Gatsby knows nothing about how to resolve these to different types I know I need to jump in and create types for each template and then I need to set 'blocks' to a union in gatsby-node.js, but when I do that I lose type inference...

```js
// gatsby-node.js
exports.createSchemaCustomization = ({ actions: { createTypes }, schema }) => {
  const typeDefs = [
    "type MarkdownRemark implements Node { frontmatter: Frontmatter }",
    schema.buildObjectType({
      name: "BlockHero",
      fields: {
        template: 'String!',
        heroField1: 'String' // intentionally not defining heroField2
      },
      interfaces: ["Node"],
    }),
    schema.buildObjectType({
      name: "BlockColumns",
      fields: {
        template: 'String!',
        columnField1: 'String!', // intentionally not defining columnField2
      },
      interfaces: ["Node"],
    }),
    schema.buildUnionType({
      name: "HeroColumnUnion",
      types: ['BlockHero', 'BlockColumns' ],
      resolveType(value) {
        if (value.template === 'block__hero') {
          return 'BlockHero';
        }
        if (value.template === 'block__columns') {
          return 'BlockColumns';
        }

        throw 'No template defined'
      }
    }),
    schema.buildObjectType({
      name: "Frontmatter",
      fields: {
        blocks: {
          type: ['HeroColumnUnion'],
        }
      },
    }),
  ]
  createTypes(typeDefs)
};
```

The problem with this is that I lose the ability for Gatsby to infer the rest of my types, so I'm left with only the fields I've defined:

```gql
query MyQuery {
  allMarkdownRemark {
    edges {
      node {
        frontmatter {
          blocks {
            ...on BlockHero {
              heroField1
            }
            ...on BlockColumns {
              columnField1
            }
          }
        }
      }
    }
  }
}
```
However the following won't work (since `heroField2` and `columnField2` aren't explicitly defined:

```gql
query MyQuery {
  allMarkdownRemark {
    edges {
      node {
        frontmatter {
          blocks {
            ...on BlockHero {
              heroField1
              heroField2
            }
            ...on BlockColumns {
              columnField1
              columnField2
            }
          }
        }
      }
    }
  }
}
```
I guess the main question I'm getting at is that in the docs I was under the impression that using buildObjectType would still allow gatsby to continue to infer the type based on data passing through it

**Comment by bernharduw** [comment]
Have you tried the @infer extension, or `extensions: { infer: true }` when using schema.buildObjectType? I had success with that in another case.

**Comment by jeffsee55** [comment]
Thanks @bernharduw - I did try that and it didn't make a difference, interestingly it doesn't look like the `buildUnionType` function supports any such config. To prove that this seems to be an issue specifically with `buildUnionType` you can do this and see type inference working properly:

```
...
schema.buildObjectType({
      name: "Frontmatter",
      fields: {
        blocks: {
          type: ['BlockColumns'], // replace HeroColumnUnion with BlockColumns to show things working as expected when just working with `buildObjectType`
        }
      },
    }),
```

**Comment by LekoArts** [comment]
cc @freiksenet @vladar @stefanprobst

**Comment by vladar** [comment]
I did some research and yeah, at the moment this is not supported. Gatsby gives up inference for this case here:

https://github.com/gatsbyjs/gatsby/blob/e65848cd3d0b826bfbedaeceecb0bd196f448725/packages/gatsby/src/schema/infer/add-inferred-fields.js#L334-L337

It will be also hard to implement. By this line Gatsby has a following inferred value for `MarkdownRemark` node type:

```
const inferredMarkdownRemarkValue = {
  'title': 'First Blog',
  'blocks': [{
    'template': 'block__hero',
    'heroField1': '## Hello world',
    'heroField2': '## Hello world',
    'columnField1': '## Hello again',
    'columnField2': '## Hello again',
  }],
};
```

As you can see Gatsby aggregates all node values into a single example object during inference. But then it is impossible to figure out which field belongs to which member of the union. 

So at the moment, the only thing we can do is add **all** inferred fields to all union members (as nullable). But then there is little point of having union type at all here.

The "proper" fix would require another pass of inference through all nodes of this type during schema building. Which will be slow.

**Comment by stonehz** [comment]
is this still an issue?

**Comment by amykapernick** [comment]
I'm having this issue now as well. Everything was working fine, and now it's decided it can't find my content field, despite it existing on most pages, but when I define it it loses all the inferred fields, despite the fact that I said `@infer` all the way down

```
exports.createSchemaCustomization = ({ actions }) => {
	const { createTypes } = actions,
		typeDefs = `
		type StoryblokEntry implements Node @infer {
			fields: Fields
		} 
		type Fields @infer {
			content: Content
		}
		type Content @infer {
			body: Body
		}
		type Body @infer {
			content: String!
			illustration: Illustration
		}
		type Illustration @infer {
			component: String!
		}
	`;

	createTypes(typeDefs);
};
```

**Comment by lagerfeuer** [comment]
@vladar instead of opening a new issue, I'm just gonna ask here: has this changed, or was it fixed? I'm facing the same issue. I'm trying to extend a source plugin that previously only returned JSON for all dynamic types. I'm trying to turn them into union types, but I can't figure out how to tell Gatsby which nested object should be which type.
